File: src/main/java/net/benwoodworth/fastcraft/core/api/dependencies/player/FcPlayer.java
package net.benwoodworth.fastcraft.core.api.dependencies.player;

import net.benwoodworth.fastcraft.core.api.dependencies.inventory.FcInventory;
import net.benwoodworth.fastcraft.core.api.dependencies.permissions.Permission;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

/**
 * A Minecraft player.
 *
 * @param <TItem> The native item type that the player carries.
 */
public interface FcPlayer<TItem> {

    /**
     * Get the player's username.
     */
    @NotNull
    String getUsername();

    /**
     * Get the player's display name.
     */
    @Nullable
    String getDisplayName();

    /**
     * Set the player's display name.
     */
    void setDisplayName(@Nullable String displayName);

    /**
     * The player's UUID.
     */
    @NotNull
    UUID getUuid();

    /**
     * Send a message to the player.
     *
     * @param message The message to send.
     */
    void sendMessage(@NotNull String message);

    /**
     * See if the player has a permission.
     *
     * @param permission The permission to check.
     */
    boolean hasPermission(@NotNull Permission permission);

    /**
     * Open an inventory.
     *
     * @param inventory The inventory to open.
     */
    void openInventory(@NotNull FcInventory<TItem> inventory);
}

File: src/main/kotlin/net/benwoodworth/fastcraft/core/dependencies/player/FcPlayer.kt
package net.benwoodworth.fastcraft.core.dependencies.player

import net.benwoodworth.fastcraft.core.dependencies.inventory.FcInventory
import net.benwoodworth.fastcraft.core.dependencies.inventory.FcItem
import net.benwoodworth.fastcraft.core.dependencies.permissions.Permission
import java.util.*

/**
 * A Minecraft player.
 *
 * @param TItem The type of item the player carries.
 */
interface FcPlayer<TItem : FcItem<*>> {

    /** The player's username. */
    val username: String

    /** The player's display name. */
    var displayName: String?

    /** The player's UUID. */
    val uuid: UUID

    /**
     * Send a message to the player.
     *
     * @param message The message to send.
     */
    fun sendMessage(message: String)

    /**
     * See if the player has a permission.
     *
     * @param permission The permission to check.
     */
    fun hasPermission(permission: Permission): Boolean

    /**
     * Open an inventory.
     *
     * @param inventory The inventory to open.
     */
    fun openInventory(inventory: FcInventory<TItem>)
}
